app/validators/program_validator.rb

```diff
@@ -1,17 +1,36 @@
 # frozen_string_literal: true
 
 class ProgramValidator < BaseValidator
-  def validate(record)
-    duplicates_exist = Program.where(['description = ? AND facility_code = ? AND id != ?',
-                                      record.description, record.facility_code, record.id]).any?
-    record.errors[:base] << non_unique_error_msg(record) if duplicates_exist
+  def self.after_import_batch_validations(failed_instances)
```

Comment:
At this point we have just entered the `after_import_batch_validations` method.
`failed_instances` contains all records that failed ActiveRecord validations.  For `Program` those are
```ruby
validates :facility_code, presence: true
validates :program_type, inclusion: { in: InstitutionProgram::PROGRAM_TYPES }
```
This is a good opportunity to add the `row` to those records' `errors` before we start adding more records to `failed_instances` with our bulk validations.  This ensures that these failed instances will display a row number in their warning messages to the end user.
```suggestion
  def self.after_import_batch_validations(failed_instances)
    failed_instances.each do |record|
      record.errors.add(:row, record.csv_row)
    end 
```

Reply:
added

Reply:
This will only add `:row` values for the records that failed on ActiveRecord::Import#import for Program CSV

Reply:
Correct, which is what `failed_instances` contains at this point in execution.

Reply:
I have moved the call into `loader.rb` and it uses `try` so that all CSV types will get the `:row` hash property for rows that do not fail validations in `after_import_validations`